### PR TITLE
Make power off tune uninterruptable

### PR DESCRIPTION
**Pull request by junwoo091400**
**Describe problem solved by this pull request**
Currently when you have a low battery warning, pressing down the power button (e.g. on ATL Mantis Edu) doesn't play the power-off tune, because it gets overridden by the low battery warning tune request published by the Commander
1. Commander was sending the "low battery warning" tune request on every single loop, thus filling up the uORB queue of the `tune_control` message.

**Describe your solution**
1. Fix the commander's `set_tune` function to only allow transmitting the tune_control uORB message request for a tune only after it has finished playing, so that it wouldn't fill up the queue excessively
2. Allow 'override' when requesting the Power button down tune in Commander

**Describe possible alternatives**
I think that a 'priority' should exist that allows override on certain tones over others, etc.

Or, the `Tunes` library that all the module / drivers actually playing the tune should integrate this prioritization in a smarter level, so that an already-playing tune wouldn't get overridden in cases like this. :thinking: 

## Discussion
I think that having this knowledge of "tune duration" in the `commander_helper` level, and limiting the rate of sending out the `tune_control` message with it's own logic is quite weird.

But since it's all connected to the uORB, it's quite tricky to liimit it's rate of publishing without the knowledge of the tune duration as well :shrug:

### CODE REVIEW

**Review by MaEtUgR — COMMENTED** [review]
I like your refactoring. It's much more readable and makes sense to me. As the next step, I think these priorities should be handled by the consumer which is the ToneAlarm driver but let's go with this step first. 👍

File: src/modules/commander/commander_helper.cpp [hunk 1]
```diff
@@ -156,36 +164,57 @@ void buzzer_deinit()
...
-		if (tune != tune_current || new_tune_duration != 0) {
-			tune_control.tune_id = tune;
-			tune_control.volume = tune_control_s::VOLUME_LEVEL_DEFAULT;
-			tune_control.tune_override = false;
-			tune_control.timestamp = hrt_absolute_time();
-			orb_publish(ORB_ID(tune_control), tune_control_pub, &tune_control);
+	const hrt_abstime current_time = hrt_absolute_time();
+	const unsigned int new_tune_duration = tune_durations[tune_id];
+	const bool current_tune_is_repeating = (tune_end == 0) ? true : false;
+	const bool new_tune_is_repeating = (new_tune_duration == 0) ? true : false;
```
Comment (MaEtUgR) [inline src/modules/commander/commander_helper.cpp:181]:
```suggestion
	const bool current_tune_is_repeating = (tune_end == 0);
	const bool new_tune_is_repeating = (new_tune_duration == 0);
```

**Review by MaEtUgR — APPROVED** [review]

**Comment by junwoo091400** [comment]
The tune_control refactor commits have been moved to a [separate PR](https://github.com/PX4/PX4-Autopilot/pull/19654), and this PR should be ready to go @MaEtUgR :+1:

File: src/modules/commander/commander_helper.h [hunk 2]
```diff
@@ -60,8 +60,24 @@ bool is_ground_rover(const vehicle_status_s &current_status);
...
+void set_tune_override(const int tune_id);
+
+/**
+ * @brief Set the new tune to play under predefined conditions
+ *
+ * Setting a new tune will only be possible for the following 3 cases:
+ *
+ * 1. Current playing tune is non-repeating and has ended
+ * 2. Current playing tune is non-repeating, but the requested tune is also non-repeating, and they are different tunes
+ * 3. Current playing tune is repeating, and the requested tune is different from teh current tune
```
Comment (MaEtUgR) [inline src/modules/commander/commander_helper.h:75]:
```suggestion
 * 3. Current playing tune is repeating, and the requested tune is different from the current tune
```

**Review by MaEtUgR — APPROVED** [review]
Yes, I still think it's a good improvement step 👍 Let's fix the typo.

**Review by MaEtUgR — APPROVED** [review]

**Pull request merged**